File: flutter_app/lib/api/chat_api.dart
import '../models/chat_message.dart';
import '../models/conversation.dart';
import 'api_client.dart';

class ChatApi {
  ChatApi(this._client);

  final ApiClient _client;

  Future<int> startConversation(int otherUserId) async {
    final data = await _client.post('/chat/start', body: {
      'other_user_id': otherUserId,
    }) as Map<String, dynamic>;
    return data['conversation_id'] as int;
  }

  Future<List<Conversation>> listConversations() async {
    final data = await _client.get('/chat/conversations') as List<dynamic>;
    return data
        .map((dynamic e) => Conversation.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  Future<List<ChatMessage>> messages({
    required int conversationId,
    int afterId = 0,
  }) async {
    final data = await _client.get('/chat/messages', query: {
      'conversation_id': '$conversationId',
      'after_id': '$afterId',
    }) as List<dynamic>;
    return data
        .map((dynamic e) => ChatMessage.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  Future<List<ChatMessage>> longPoll({
    required int conversationId,
    int afterId = 0,
    int timeout = 30,
  }) async {
    final data = await _client.get('/chat/longpoll', query: {
      'conversation_id': '$conversationId',
      'after_id': '$afterId',
      'timeout': '$timeout',
    }) as Map<String, dynamic>;
    final messages = data['messages'] as List<dynamic>? ?? const [];
    return messages
        .map((dynamic e) => ChatMessage.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  Future<ChatMessage> send({
    required int conversationId,
    required String content,
  }) async {
    final data = await _client.post('/chat/send', body: {
      'conversation_id': conversationId,
      'content': content,
      'message_type': 'text',
    }) as Map<String, dynamic>;
    return ChatMessage.fromJson(data);
  }

  Future<void> markRead(int conversationId) async {
    await _client.post('/chat/mark-read', body: {
      'conversation_id': conversationId,
    });
  }
}

File: frontend/src/api/chatApi.ts
import { ChatMessage, parseChatMessage } from '../models/chatMessage';
import { Conversation, parseConversation } from '../models/conversation';
import { asArray, asRecord, asInt } from '../models/parse';
import { ApiClient } from './apiClient';

export class ChatApi {
  constructor(private readonly client: ApiClient) {}

  async startConversation(otherUserId: number): Promise<number> {
    const data = asRecord(
      await this.client.post('/chat/start', {
        other_user_id: otherUserId,
      }),
    );
    return asInt(data.conversation_id);
  }

  async listConversations(): Promise<Conversation[]> {
    const data = asArray(await this.client.get('/chat/conversations'));
    return data.map(parseConversation);
  }

  async messages(conversationId: number, afterId = 0): Promise<ChatMessage[]> {
    const data = asArray(
      await this.client.get('/chat/messages', {
        conversation_id: String(conversationId),
        after_id: String(afterId),
      }),
    );
    return data.map(parseChatMessage);
  }

  async longPoll(conversationId: number, afterId = 0, timeout = 30): Promise<ChatMessage[]> {
    const data = asRecord(
      await this.client.get('/chat/longpoll', {
        conversation_id: String(conversationId),
        after_id: String(afterId),
        timeout: String(timeout),
      }),
    );
    return asArray(data.messages).map(parseChatMessage);
  }

  async send(conversationId: number, content: string): Promise<ChatMessage> {
    const data = await this.client.post('/chat/send', {
      conversation_id: conversationId,
      content,
      message_type: 'text',
    });
    return parseChatMessage(data);
  }

  async markRead(conversationId: number): Promise<void> {
    await this.client.post('/chat/mark-read', {
      conversation_id: conversationId,
    });
  }
}
